### DISCUSSION TITLE
Configure a maximum lifetime for pooled connections

### DISCUSSION THREAD

**Discussion by jhominal** [encode/httpx#3177] [Ideas]
I think it would be nice for `httpx` (and `httpcore`) if it were possible to set the maximum lifetime of a keepalive connection.

This is different from the `keepalive_expiry` parameter, in that:
 - `keepalive_expiry`: starts once a given connection becomes idle, corresponds to `libcurl`'s [`CURLOPT_MAXAGE_CONN`](https://curl.se/libcurl/c/CURLOPT_MAXAGE_CONN.html);
 - `keepalive_max_lifetime` (name to be workshopped): starts as soon as the connection is created, corresponds to `libcurl`'s [`CURLOPT_MAXLIFETIME_CONN`](https://curl.se/libcurl/c/CURLOPT_MAXLIFETIME_CONN.html);

This would probably be modeled as a new field in `httpx.Limits` and `httpcore.ConnectionPool`.

**Comment by lovelydinosaur** [comment]
That's neatly coherent with the rest of the API there, yup.
Any pointers on default behaviors from other agents here?
Eg. Do browsers have a max age on persistence for HTTP connections?

**Reply by jhominal** [reply]
From what I can see:
 * `libcurl` disables maximum lifetime by default;
 * from chromium source code, I think that there are idle timeouts (equivalent to `keepalive_expiry`), but there does not seem to be a total lifetime timeout;
 * gecko seems also to only have a keep alive with an idle timeout mechanism, mostly set (on HTTP 1.1) by parsing a timeout parameter in `Keep-Alive`;
 * `urllib3` does not seem to have mechanisms to limit connection reuse after they are open;

In short:

 * Most networking libraries do not support such a feature (mostly, I have only seen it in `libcurl`);
 * Even for `libcurl`, this mechanism is disabled by default;
